soroban-sdk/src/address.rs

```diff
@@ -1,13 +1,208 @@
-use crate::{accounts::AccountId, contracttype, BytesN};
-
-/// Address is all the types that can be the invoker of a contract.
-///
-/// The Address type is a contracttype and transmitted to the host like an enum,
-/// however it is not exported into contract specs because it has its own spec
-/// definition type dedicated to it.
-#[contracttype(crate_path = "crate", lib = "soroban_sdk", export = false)]
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
-pub enum Address {
-    Account(AccountId),
-    Contract(BytesN<32>),
+use core::{cmp::Ordering, fmt::Debug};
+
+use super::{
+    env::internal::{Env as _, EnvBase as _},
+    xdr::ScObjectType,
+    ConversionError, Env, Object, RawVal, TryFromVal,
+};
+
+#[cfg(not(target_family = "wasm"))]
+use crate::env::internal::xdr::ScVal;
+use crate::{unwrap::UnwrapInfallible, Vec};
+
+/// Address is a universal opaque identifier to use in contracts.
+///
+/// Address can be used as an input argument (for example, to identify the
+/// payment recipient), as a data key (for example, to store the balance), as
+/// the authentication & authorization source (for example, to authorize the
+/// token transfer) etc.
+///
+/// See `require_auth` documentation for more details on using Address for
+/// authorization.
+///
+/// Internally, Address may represent a Stellar account or a contract. Contract
+/// address may be used to identify the account contracts - special contracts
+/// that allow customizing authentication logic and adding custom authorization
+/// rules.
+///
+/// In tests Addresses should be generated via `Address::random()`.
+#[derive(Clone)]
+pub struct Address {
+    env: Env,
+    obj: Object,
+}
+
+impl Debug for Address {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        #[cfg(target_family = "wasm")]
+        write!(f, "Address(..)")?;
+        #[cfg(not(target_family = "wasm"))]
+        {
+            use crate::env::internal::xdr;
+            use stellar_strkey::StrkeyPublicKeyEd25519;
+            let sc_val = ScVal::try_from(self).map_err(|_| core::fmt::Error)?;
+            if let ScVal::Object(Some(xdr::ScObject::Address(addr))) = sc_val {
+                match addr {
+                    xdr::ScAddress::Account(account_id) => {
+                        let xdr::AccountId(xdr::PublicKey::PublicKeyTypeEd25519(xdr::Uint256(
+                            ed25519,
+                        ))) = account_id;
+                        let strkey = StrkeyPublicKeyEd25519(ed25519);
+                        write!(f, "AccountId({})", strkey.to_string())?;
+                    }
+                    xdr::ScAddress::Contract(contract_id) => {
+                        // TODO: we should use contract strkey here when it's
+                        // supported.
+                        write!(f, "Contract({:?})", contract_id.0)?;
```

Comment:
It's supported: http://docs.rs/stellar-strkey/latest/stellar_strkey/struct.Contract.html